packages/eslint-plugin-internal/src/rules/prefer-ast-types-enum.ts

```diff
@@ -0,0 +1,70 @@
+import {
+  AST_NODE_TYPES,
+  AST_TOKEN_TYPES,
+  ESLintUtils,
+  TSESTree,
+} from '@typescript-eslint/experimental-utils';
+
+const isStringLiteral = (
+  node: TSESTree.Literal,
+): node is TSESTree.StringLiteral => typeof node.value === 'string';
+
+export = ESLintUtils.RuleCreator(name => name)({
+  name: __filename,
+  meta: {
+    type: 'problem',
+    docs: {
+      category: 'Best Practices',
+      recommended: 'error',
+      description:
+        'Ensures consistent usage of AST_NODE_TYPES & AST_TOKEN_TYPES enums.',
+    },
+    messages: {
+      preferEnum: 'Prefer {{ enumName }}.{{ literal }} over raw literal',
+    },
+    fixable: 'code',
+    schema: [],
+  },
+  defaultOptions: [],
+  create(context) {
+    const report = (
+      enumName: 'AST_NODE_TYPES' | 'AST_TOKEN_TYPES',
+      literal: TSESTree.StringLiteral,
+    ): void =>
+      context.report({
+        data: { enumName, literal: literal.value },
+        messageId: 'preferEnum',
+        node: literal,
+        fix: fixer =>
+          fixer.replaceText(literal, `${enumName}.${literal.value}`),
+      });
+
+    return {
+      Literal(node: TSESTree.Literal): void {
+        if (
+          node.parent?.type === AST_NODE_TYPES.TSEnumMember &&
+          node.parent.parent?.type === AST_NODE_TYPES.TSEnumDeclaration &&
+          ['AST_NODE_TYPES', 'AST_TOKEN_TYPES'].includes(
+            node.parent.parent.id.name,
+          )
+        ) {
+          return;
+        }
+
+        if (!isStringLiteral(node)) {
+          return;
+        }
+
+        const value = node.value;
+
+        if (value in AST_NODE_TYPES) {
```

Comment:
`constructor` shouldn't match here
This is a bug that I didn't think about.

These two checks should instead be :
```suggestion
        if (AST_NODE_TYPES.hasOwnProperty(value)) {
```

Reply:
ugh you're completely right - we had the same issue in `eslint-plugin-jest` 🤦‍♂

Reply:
I'll add a test for that once I've finished re-running `--fix`

Reply:
I love how we both are "javascript experts", and yet we both still forgot about such a noobish trap. 😢 

![FxsV0ja](https://user-images.githubusercontent.com/7462525/73104764-0d7fd580-3eac-11ea-8009-4eda1eb8b83c.gif)